src/libopensc/card-eoi.c

```diff
@@ -0,0 +1,544 @@
+	{ NULL, NULL, NULL, 0, 0, NULL }
+};
+
+/*
+ * CAN is stored encrypted in a file that (looks like) is pointed to by 'Card CAN' PIN object.
+ * eoi_decrypt_can() decrypts CAN from it's encrypted form
+ */
+
+static void rol(u8 *to, const u8 *from)
+{
+	int i;
+	u8 b = from[0] & 0x80;
+	for (i = 15; i >= 0; i--) {
+		u8 bo = b;
+		b = from[i] & 0x80;
+		to[i] = (from[i] << 1) | (bo ? 1 : 0);
+		if ((i == 15) && bo)
+			to[i] = (to[i] ^ 0x87) | 1;
+	}
+}
+
+static int aes256_ecb_encrypt(const u8 *key, const u8 *input, const u8 *output)
+{
+	EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
+	int r = 0, pos, len = pos = AES_BLOCK_SIZE;
+	if (!ctx)
+		goto err;
+	if (!EVP_EncryptInit(ctx, EVP_aes_256_ecb(), key, NULL))
+		goto err;
+	if (!EVP_EncryptUpdate(ctx, (u8 *)output, &pos, input, len))
+		goto err;
+	len -= pos;
+	/*
+	 * If EVP_EncryptFinal() is called with len = 0, at least with OpenSSL 3.0.2,
+	 * app crashes with ***stack smashing detected***
+	 */
+	if (len != 0) {
+		if (!EVP_EncryptFinal(ctx, (u8 *)(output + pos), &len))
+			goto err;
+	}
+	r = 1;
+err:
+	if (ctx)
+		EVP_CIPHER_CTX_free(ctx);
+	return r;
+}
+
+static int aes256_ecb_decrypt(const u8 *key, const u8 *input, const u8 *output)
+{
+	EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
+	int r = 0, pos, len = pos = AES_BLOCK_SIZE;
+	if (!ctx)
+		goto err;
+	if (!EVP_DecryptInit(ctx, EVP_aes_256_ecb(), key, NULL))
+		goto err;
+	/* Disable padding, otherwise it will fail to decrypt non-padded inputs */
+	if (!EVP_CIPHER_CTX_set_padding(ctx, 0))
+		goto err;
+	if (!EVP_DecryptUpdate(ctx, (u8 *)output, &pos, input, len))
+		goto err;
+	len -= pos;
+	/* Better safe than sorry */
+	if (len != 0) {
+		if (!EVP_DecryptFinal(ctx, (u8 *)(output + pos), &len))
+			goto err;
+	}
+	r = 1;
+err:
+	if (ctx)
+		EVP_CIPHER_CTX_free(ctx);
+	return r;
+}
+
+/*
+ * CAN decrypt magic...
+ */
+static int get_can_key(const u8 *key, const u8 round, const u8 *input, u8 *output)
+{
+	size_t i;
+	u8 tmp[3][AES_BLOCK_SIZE];
+	memset(tmp[0], 0, AES_BLOCK_SIZE);
+	if (!aes256_ecb_encrypt(key, tmp[0], tmp[0]))
+		return 0;
+	rol(tmp[1], tmp[0]);
+	rol(tmp[0], tmp[1]);
+	memset(tmp[1], 0, AES_BLOCK_SIZE);
+	tmp[1][11] = 4;
+	tmp[1][13] = 1;
+	tmp[1][15] = round;
+	if (!aes256_ecb_encrypt(key, tmp[1], tmp[2]))
+		return 0;
+	memset(tmp[1], 0, AES_BLOCK_SIZE);
+	memcpy(tmp[1], &input[AES_BLOCK_SIZE], 8);
+	tmp[1][8] = 0x80;
+	for (i = 0; i < AES_BLOCK_SIZE; i++)
+		tmp[0][i] = tmp[0][i] ^ tmp[1][i] ^ tmp[2][i];
+	if (!aes256_ecb_encrypt(key, tmp[0], output))
+		return 0;
+	return 1;
+}
+
+static int eoi_decrypt_can(struct sc_pkcs15_u8 *enc_can, char *can) {
+	/* Magic key that is used to decrypt CAN */
+	const u8 magic_key[32] = {0xC8, 0x12, 0x0F, 0xD8, 0x21, 0x20, 0x1F, 0x77, 0xF1, 0x83, 0x9D, 0xD8, 0x86, 0xB0, 0x5C, 0xF2, 0x4F, 0x7E, 0x52, 0x66, 0xE5, 0x87, 0x89, 0x2B, 0xF4, 0xC5, 0xE5, 0x4C, 0x54, 0xA1, 0x55, 0x30};
+	u8 can_key[32] = { 0 };
```

Comment:
Maybe replace 32 with `2*AES_BLOCK_SIZE`?

Reply:
I disagree. 32 in this case is the length of the AES key and has nothing to do with the AES block size.
I've added `#define AES256_KEY_LEN 32` to make it more obvious...

Reply:
thank you, my fault